docs/custom_promise_types/rss.py

```diff
@@ -0,0 +1,165 @@
+        items = self._get_items(resource)
+        if not items:
+            return Result.NOT_KEPT
+
+        # pick item based on optional select attribute (defaults to newest)
+        item = self._pick_item(items, attributes)
+
+        # write selected item to promiser file
+        result = self._write_promiser(item, promiser)
+
+        return result
+
+
+    def _get_resource(self, path):
+        if self._is_url(path):
+            # fetch from url
+            self.log_verbose(f"Fetching feed from url '{path}'")
+            response = requests.get(path)
+            if response.ok:
+                return response.content
+            self.log_error(f"Failed to fetch feed from url '{path}'': status code '{response.status_code}'")
+            return None
+
+        # fetch from file
+        try:
+            self.log_verbose(f"Reading feed from file '{path}'")
+            with open(path, 'r') as f:
+                resource = f.read()
+                return resource
+        except Exception as e:
+            self.log_error(f"Failed to open file '{path}' for reading: {e}")
+            return None
+
+
+    def _get_items(self, res):
+        # extract descriptions in /channel/item
+        try:
+            self.log_verbose("Parsing feed")
+            items = []
+            root = ET.fromstring(res)
+            for item in root.findall('./channel/item'):
+                for child in item:
+                    if child.tag == 'description':
+                        items.append(child.text)
+            return items
+        except Exception as e:
+            self.log_error(f"Failed to parse feed: {e}")
+            return None
+
+
+    def _pick_item(self, items, attributes):
+        # Pick newest item as default
+        item = items[0]
+
+        # Select item from feed
+        if "select" in attributes:
+            select = attributes['select']
+            if select == 'random':
+                self.log_verbose("Selecting random item from feed")
+                item = random.choice(items)
+            elif select == 'oldest':
+                self.log_verbose("Selecting oldest item from feed")
+                item = items[- 1]
+            else:
+                self.log_verbose("Selecting newest item from feed")
+        else:
+            self.log_verbose("Selecting newest item as default")
+        return item
+
+
+    def _write_promiser(self, item, promiser):
+        file_exist = os.path.isfile(promiser)
+
+        if file_exist:
+            try:
+                with open(promiser, 'r') as f:
+                    if f.read() == item:
+                        self.log_verbose(f"File '{promiser}' exists and is up to date, no changes needed")
+                        return Result.KEPT
+            except Exception as e:
+                self.log_error(f"Failed to open file '{promiser}' for reading: {e}")
+                return Result.NOT_KEPT
+
+        try:
+            with open(promiser, 'w') as f:
```

Comment:
I am guessing that when you get data from XML / requests, it is utf-8 encoded. Most places in python, that is the default, however:

> Normally, files are opened in text mode, that means, you read and write strings from and to the file, which are encoded in a specific encoding. If encoding is not specified, the default is platform dependent (see [`open()`](https://docs.python.org/3/library/functions.html#open)). 'b' appended to the mode opens the file in binary mode: now the data is read and written in the form of bytes objects. This mode should be used for all files that don’t contain text.

https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files

It seems like on the platform @nickanderson is testing, the default encoding chosen here is ascii, not utf-8. Thus, if your input is utf-8 and includes non-ascii characters, and you try to write it as ascii, you get exceptions. You can force utf-8 encoding of the written data:

```suggestion
            with open(promiser, 'w', encoding='utf-8') as f:
```

Reply:
In addition to patching here where open for writing, I patched where it opened for reading and my ouptut is much improved.

```
[root@demohub ~]# cat /tmp/word-of-the-day-random.txt
OED Word of the Day: sari-sari store, n. In the Philippines: a small neighbourhood store selling a variety of goods.[root@demohub ~]# cat cf-agent -KIf /var/cfengine/inputs/services/custom_promise_types/rss/policy/main.cf     info: File '/tmp/ISS-random.txt' exists but contents differ, updating content
    info: File '/tmp/word-of-the-day-random.txt' exists but contents differ, updating content
   error: Failed to open file '/tmp/rss.xml' for reading: 'ascii' codec can't decode byte 0xe2 in position 3168: ordinal not in range(128)
    info: File '/tmp/word-of-the-day-random.txt' exists but contents differ, updating content
   error: Failed to open file '/tmp/rss.xml' for reading: 'ascii' codec can't decode byte 0xe2 in position 3168: ordinal not in range(128)
    info: File '/tmp/ISS-random.txt' exists but contents differ, updating content
   error: Failed to open file '/tmp/rss.xml' for reading: 'ascii' codec can't decode byte 0xe2 in position 3168: ordinal not in range(128)
   error: Method 'example_rss_promise' failed in some repairs
[root@demohub ~]# cat /tmp/word-of-the-day-random.txt
OED Word of the Day: ice master, n. A naval officer or pilot who has special experience in navigating vessels in icy waterways.
```